### PR TITLE
Secp256k1 bc verify using embedded signature

### PR DESCRIPTION
**Pull request by rschust**
#161 The voting history test can now verify SECP256K1 signatures with a recoverable embedded public key.
Signature serialization now includes the recovery id in addition to the signature.
SECP256K1 verification now completely relies on the embedded signature (keypair.h).
Added coincurve as a pip dependency.
There are 2 functions on votinghistory.py that can verify SECP256K1 BC signatures:

- verify_recover_secp256k1_bc verifies the signature using the coincurve API, as well as a single native call, since the API does not provide a clean verification method (should create a PR on the coincurve repository to add this)

- verify_recover_secp256k1_bc_native which uses the native API directly, and can be used for reference or in case the imported library changes in a way that breaks

### CODE REVIEW

File: src/tls/keypair.h [hunk 1]
```diff
@@ -54,10 +54,47 @@ namespace tls
...
+    secp256k1_context* ctx, const uint8_t* signature, const uint8_t* hash)
+  {
+    secp256k1_pubkey public_key;
+    secp256k1_ecdsa_recoverable_signature sig;
+
+    if (
+      secp256k1_ecdsa_recoverable_signature_parse_compact(
+        ctx, &sig, signature, signature[REC_ID_IDX]) != 1)
+    {
+      LOG_INFO_FMT(
```
Comment (jumaffre) [inline src/tls/keypair.h]: Here and in error cases in this function: `LOG_FAIL_FMT`

File: tests/votinghistory.py [hunk 2]
```diff
@@ -7,19 +7,75 @@
...
+from coincurve.context import GLOBAL_CONTEXT
+from coincurve.ecdsa import deserialize_recoverable, recover
+from coincurve.utils import bytes_to_int, sha256
 
 import cryptography.x509
 from cryptography.hazmat.primitives.asymmetric import ec
 from cryptography.hazmat.backends import default_backend
 
+# This function calls the native API and does not rely on the
+# imnport library's implementation. Though not being used by
```
Comment (jumaffre) [inline tests/votinghistory.py]: Small type on "import"

File: src/tls/keypair.h (hunk 1 above)
Comment (rschust) [reply src/tls/keypair.h]: Not sure this should be logged as an error. If we get a lot of invalid signed requests, it'll spam our logs, and we never really logged these failures before.

Comment (jumaffre) [reply src/tls/keypair.h]: True. Perhaps `LOG_DEBUG_FMT` then? info will spam the logs just as much as fail in release builds.

File: src/tls/keypair.h [hunk 3]
```diff
@@ -54,10 +54,47 @@ namespace tls
 #elif CURVE_CHOICE_SECP256K1_MBEDTLS || CURVE_CHOICE_SECP256K1_BITCOIN
   static constexpr mbedtls_md_type_t MD_TYPE = MBEDTLS_MD_SHA256;
   static constexpr size_t MD_SIZE = 256 / 8;
+  static constexpr size_t REC_ID_IDX = 64;
 #endif
 
   using Hash = std::array<uint8_t, MD_SIZE>;
 
+#if CURVE_CHOICE_SECP256K1_BITCOIN
```
Comment (achamayou) [inline src/tls/keypair.h]: Do we need the ifdef here?

File: tests/votinghistory.py [hunk 4]
```diff
@@ -7,19 +7,75 @@
...
+from coincurve.context import GLOBAL_CONTEXT
+from coincurve.ecdsa import deserialize_recoverable, recover
+from coincurve.utils import bytes_to_int, sha256
 
 import cryptography.x509
 from cryptography.hazmat.primitives.asymmetric import ec
 from cryptography.hazmat.backends import default_backend
 
+# This function calls the native API and does not rely on the
+# import library's implementation. Though not being used by
```
Comment (achamayou) [inline tests/votinghistory.py]: imported?

File: src/tls/keypair.h (hunk 3 above)
Comment (rschust) [reply src/tls/keypair.h]: Not really, will be removed

**Pull request merged**